pootle/apps/virtualfolder/receivers.py

```diff
@@ -19,85 +20,67 @@ def update_vfolder_tree(vf, store):
     """
     # Create missing VirtualFolderTreeItem tree structure for affected Stores
     # after adding or unobsoleting Units.
-    vfolder_treeitem, created = (
-        VirtualFolderTreeItem.objects.get_or_create(
-            directory=store.parent, vfolder=vf))
-
+    created = False
+    try:
+        vfolder_treeitem = (
+            VirtualFolderTreeItem.objects.get(
+                directory=store.parent, vfolder=vf))
+    except VirtualFolderTreeItem.DoesNotExist:
+        parts = split_pootle_path(store.parent.pootle_path)
+        path_parts = ["", parts[0], parts[1], vf.name]
+        if parts[2]:
+            path_parts.append(parts[2].strip("/"))
+        path_parts.append(parts[3])
+        pootle_path = "/".join(path_parts)
+        vfolder_treeitem = (
+            VirtualFolderTreeItem.objects.create(
+                pootle_path=pootle_path,
+                directory=store.parent,
+                vfolder=vf))
+        created = True
     if not created:
         # The VirtualFolderTreeItem already existed, so
         # calculate again the stats up to the root.
         vfolder_treeitem.update_all_cache()
 
 
-def add_store_to_vfolders(store):
-    """For a given Unit check for membership of any VirtualFolders
-    """
-    pootle_path = store.pootle_path
-
-    for vf in VirtualFolder.objects.iterator():
-        store_added = False
-        for location in vf.all_locations:
-            if not pootle_path.startswith(location):
-                continue
-
-            for filename in vf.filter_rules.split(","):
-                if pootle_path == "".join([location, filename]):
-                    vf.stores.add(store)
-                    store_added = True
-                    break
-
-            if store_added:
-                break
-
-        if store_added:
-            update_vfolder_tree(vf, store)
+@receiver(post_save, sender=Store)
+def handle_store_save(sender, instance, created, **kwargs):
+    if not created:
+        return
+    project = instance.translation_project.project
+    language = instance.translation_project.language
+
+    # limit to vfolders that could be possibly related
+    vfolders = (
+        project.vfolders.filter(all_languages=True)
+        | language.vfolders.filter(all_projects=True)
+        | project.vfolders.filter(languages=language)
+        | VirtualFolder.objects.filter(
+            all_languages=True, all_projects=True))
+    for vf in vfolders:
+        added = vf.path_matcher.add_store_if_matching(instance)
+        if added:
+            update_vfolder_tree(vf, instance)
+            instance.set_priority()
```

Comment:
on that note @unho we could probs do with some code to set priority for multiple stores at once